### PR TITLE
listListenerMatching optimization

### PR DESCRIPTION
**Pull request by andrewluetgers**
I closed a prior PR because it was hopelessly messy. Hopefully this will make it easier to evaluate this change.

## listListenerMatching optimization
This optimization was the result of experimenting with a custom omniscient mixin that allows you to do reference observer based updating of components directly as opposed to the usual top down approach. https://github.com/omniscientjs/omniscient/issues/93#issuecomment-84812169
I was able to achieve 80 fps using what would be the normal approach with immstruct, with the omniscient mixin I was able to get 100 fps and with this change to the underlying datastructure used by listListenerMatching I was able to achieve 170 fps. As a reference the fastest this app can operate on my machine is at 200 fps using vanilla js. See the above link for running code samples.

Currently the only test that is failing is 'should allow values as key paths.' This implementation expects to be able to turn keyPaths into keys in a plain JS object for the tree of observers so may be incompatible with values as key paths if those "values" cannot be easily turned into a string.

### CODE REVIEW

File: src/structure.js [hunk 1]
```diff
@@ -253,10 +252,12 @@ Structure.prototype.reference = function (path) {
      * @module reference.cursor
      * @returns {Cursor} Immutable.js cursor
      */
-    cursor: function (subPath) {
-      subPath = valToKeyPath(subPath);
-      if (subPath) return cursor.cursor(subPath);
-      return cursor;
+    cursor: function (path) {
```
Comment (mikaelbr) [inline src/structure.js]: Why do this need to change? Removing `valToKeyPath` will cause cursors to behave differently through references and normal cursors.

File: src/structure.js [hunk 2]
```diff
@@ -487,21 +492,31 @@ function hasIn(cursor, path) {
...
-  if (!path || !path.length) return topLevel;
-  return [topLevel].concat(path).join('|');
+function listListenerMatching (listeners, path) {
+  var pathArray = [],
+      matches = [];
+
+  path = path || [];
+
+  path.forEach(function(pathPart) {
+    pathArray = pathArray.concat(pathPart);
```
Comment (mikaelbr) [inline src/structure.js]: What if pathPart here is a non-string? e.g. a complex structure (object, immutable structure, etc).

File: src/structure.js [hunk 3]
```diff
@@ -487,21 +492,31 @@ function hasIn(cursor, path) {
...
 
-function listListenerMatching (listeners, basePath) {
-  var newListeners = [];
-  for (var key in listeners) {
-    if (!listeners.hasOwnProperty(key)) continue;
-    if (basePath.indexOf(key) !== 0) continue;
-    newListeners.push(listeners[key]);
+function getListenerNs (listeners, path) {
+  path = path || [];
+  path = path.concat("__listeners__");
```
Comment (mikaelbr) [inline src/structure.js]: What is this needed for? Is this internals that should be in `deepGet`?

File: src/structure.js [hunk 4]
```diff
@@ -169,13 +169,19 @@ Structure.prototype.reference = function (path) {
...
+
+  var self = this,
+      listenerNs = getListenerNs(self._pathListeners, path),
+      cursor = this.cursor(path),
+      changeListener = function() {
+        cursor = self.cursor(path);
+      };
+
+  if (!listenerNs[0]) {
+    listenerNs.push(changeListener);
```
Comment (mikaelbr) [inline src/structure.js]: It seems to me, that if you has on `getListenerNs(self._pathListeners, path)` you can get a object literal in return, not an array. This means `.push` will be undefined? As `self._pathListeners` (object) is passed into `utils.deepGet`, which returns the passed into value (which is `self._pathListeners`). And the value from `utils.deepGet` will get returned from `getListenerNs`. So `listenerNs` will be an object and `({}).push => undefined`, no?

File: src/structure.js (hunk 1 above)
Comment (andrewluetgers) [inline src/structure.js]: yes that is not needed now and fixes the failing test

File: src/structure.js (hunk 2 above)
Comment (andrewluetgers) [inline src/structure.js]: The same problem applies to the original code the only difference is that there is an implied toString call placed on all parts concat'd together in the original code because of the .join, I could simply call toString on each pathPart but for plain JS objects you will get the generic '[object Object]' string and I'm not sure what you get for the other supported keys.

File: src/structure.js (hunk 3 above)
Comment (andrewluetgers) [inline src/structure.js]: If you run the following code and look at the log you see the tree uses keys to store path structure, but also needs a place to put the callbacks so this is how I'm trying to do that and avoid any collisions.
```javascript
var struct = immstruct({foo: {bar: {baz: 1}}});
struct.reference(['foo', 'bar']).observe('swap', function() {console.log(arguments)});
struct.reference('foo').observe('swap', function() {console.log(arguments)});
console.log(struct._pathListeners);
```

File: src/structure.js (hunk 4 above)
Comment (andrewluetgers) [inline src/structure.js]: Right now this whole thing was built without consideration of using anything other than key paths that look like ['foo', 'bar', 'baz'], an array of strings. getListenerNs calls concat on the provided path, adding in the '__listeners__'  path, this array is passed to deepGet, if it has already been added it would be an array because if it is not added we set it as an array. If you're worried about potential collisions with users using that string in their code you can use some generated random string of sufficient length instead.

**Comment by andrewluetgers** [comment]
@mikaelbr here is a Map implementation with deepSet and deepGet demonstrating a data-structure that could be used in place of this current implementation using basic JS objects with string keys. It also shows how the data-structure would store the callbacks and how you could add and retrieve them.
https://jsfiddle.net/aq93v4g7/3/

**Comment by Dashed** [comment]
Looks like something similar to `deepSet`/`deepGet` is being incubated in lodash as `_.get` and `_.set`: https://github.com/lodash/lodash/commit/2edcc893034936cccb94264783453b3d6bceff45

Might be good to use that in the future, as alternative implementations, when it gets released.

**Comment by andrewluetgers** [comment]
@Dashed not exactly, this is a very specific use case and not used anywhere else in the codebase. I contributed to [lodash-deep](https://github.com/marklagendijk/lodash-deep) early on and am pretty familiar with the issues involved. Fundamentally you can cut a lot of corners in this case as it is an optimization and not a general use utility there should be no problem with that.

**Comment by Dashed** [comment]
Fair points. But there's no harm in outsourcing these utilities to eventual modules:
https://npm.im/lodash.get
https://npm.im/lodash.set

I suggest this because those API methods may have the necessary optimizations you don't need to worry about, nor maintain at a higher-level library like immstruct.

**Comment by andrewluetgers** [comment]
This is the thing, there is not going to be a deepGet, or deepSet implementation in lodash that works with a Map data-structure (takes any value for keys, uses Map.get, Map.set), it will work with native js objects and arrays (strings for keys, native object access and mutation syntax). Similarly I also could write this optimization with the Map implementation in immutable.js that would obviate the use of my custom Map but that would be overkill for a performance optimization I suspect it would use more memory and be slower. Of course only testing both will tell you that but at either rate you need some code to take a keyPath into a Map and use it to set and get values to make it all work.

https://gist.github.com/andrewluetgers/7f0f043ea2224c315ad0#file-map-js-L37-L67

**Comment by andrewluetgers** [comment]
when lodash releases a deep get, deep set I'll certainly look at it for some possible improvements but if I want to use their code I'll have to fork it to support working with a tree of Maps. There is already a [lodash-deep](https://github.com/marklagendijk/lodash-deep) which I have actually contributed to. I could have also used that but again it would be a fork and there is much complexity there because it is intended as a general use utility.

**Comment by andrewluetgers** [comment]
@mikaelbr I have merged in the Map changes so now this change supports arbitrary types for keys. I think this still needs more tests but would like your feedback. Performance-wise there is no discernible impact to switching to the new Map structure. http://plnkr.co/edit/TRh18pmklgOF8cbunYbj?p=preview

**Comment by mikaelbr** [comment]
I was experimenting with maps yesterday my self, but since we have access to Immutable.js I thought it made more sense to use Immutabe.Map. I had some kinks I needed to work out, though. But maybe it makes sense to use Immutable.Map? Also, while I was looking into the source code yesterday, I fpund thee code a bit hard to follow and I'm not the biggest fan of mutation and side-effect in the helper functions (match listeners etc).

**Comment by andrewluetgers** [comment]
I'm not sure why you need immutability now to back the storage of callbacks, was that a concern before when you were using arrays? I think the two (needing a Map vs. needing immutability) are getting conflated.

**Comment by mikaelbr** [comment]
I don't say we need immutability, but we already have a Map implementation available - and probably don't need [a custom implementation](https://github.com/andrewluetgers/immstruct/commit/2fab8fc4fe26ea306559347dd002a45607d81cdc#diff-2b4ca49d4bb0a774c4d4c1672d7aa781R10)

**Comment by andrewluetgers** [comment]
Fair enough, any suggestions on how to make the code easier to follow? Also what are the areas of mutation and side-effect that are of concern?

**Comment by Dashed** [comment]
I glanced over the rework using the Map implementation. I noticed that appending `__listeners__` to path doesn't actually avoid collisions:

```js
var _pathListeners = Map();
getListenerNs(_pathListeners, ['foo']).push(1);
getListenerNs(_pathListeners, ['foo', '__listeners__']); // => error
```

A compromise may be to make `__listeners__` configurable via Structure options.
Or, if you don't want an official API for it, make it a property `structure._listenersKey = '__listeners__'` for advanced users.

**Comment by andrewluetgers** [comment]
@Dashed yes, it is not perfect, I think it would be good to throw friendly error message if we see a key that matches the listeners key and make it configurable.

**Comment by andrewluetgers** [comment]
@Dashed I just realized now that it uses a Map the listeners key can be a unique sentinel value ala
  var listenersSentinel = {};
this would fully prevent any collisions :-)

**Comment by Dashed** [comment]
That works too :+1:

File: src/utils.js [hunk 5]
```diff
@@ -4,3 +4,81 @@ module.exports.generateRandomKey = function (len) {
...
+  },
+
+  indexOf: function(key) {
+    var i = this._keys.indexOf(key);
+    return i === -1 ? this._keys.length : i;
+  },
+
+  deepGet: function(keyPath) {
+    var current = this,
+        keyPath = keyPath || [],
```
Comment (Dashed) [inline src/utils.js]: A nitpick. Redefining `keyPath`.

File: src/utils.js [hunk 6]
```diff
@@ -4,3 +4,81 @@ module.exports.generateRandomKey = function (len) {
...
+        break;
+      }
+    }
+
+    return current;
+  },
+
+  deepSet: function(keyPath, value) {
+    var current = this,
+        keyPath = keyPath || [],
```
Comment (Dashed) [inline src/utils.js]: Also a nitpick. Redefining `keyPath`.

File: src/structure.js [hunk 7]
```diff
@@ -267,8 +264,7 @@ Structure.prototype.reference = function (path) {
      * @returns {Void}
      */
     unobserveAll: function () {
-      removeAllListenersBut(self, pathId, referenceListeners, changeListener);
-      referenceListeners = [changeListener];
+      removeAllListenersBut(listenerNs, changeListener);
```
Comment (Dashed) [inline src/structure.js]: IIRC, the original API for `unobserveAll` only removed observers added through a cursor reference. This removes listeners added by other cursor references of the same keypath.

File: src/structure.js [hunk 8]
```diff
@@ -280,8 +276,7 @@ Structure.prototype.reference = function (path) {
      * @returns {Void}
      */
     destroy: function () {
-      removeAllListenersBut(self, pathId, referenceListeners);
-      referenceListeners = void 0;
+      removeAllListenersBut(listenerNs);
```
Comment (Dashed) [inline src/structure.js]: Similar to https://github.com/omniscientjs/immstruct/pull/56/files#r28195201. This destroys all observers attached by other cursor references of the same keypath; not only its own.

File: src/utils.js [hunk 9]
```diff
@@ -4,3 +4,81 @@ module.exports.generateRandomKey = function (len) {
...
+        _cur;
+
+    keyPath.forEach(function(key, index) {
+      if (index + 1 === keyPath.length) {
+        current.set(key, value);
+      } else {
+        _cur = current.get(key);
+        if (!_cur) {
```
Comment (Dashed) [inline src/utils.js]: This wouldn't really matter in this specific use case. But rather than test for falsey, probably opt for somehing like: 
`if ((_cur = current.get(key, NOT_SET)) === NOT_SET) {`

Basing on `Map.get` http://facebook.github.io/immutable-js/docs/#/Map/get

**Comment by Dashed** [comment]
@andrewluetgers You should probably put this map implementation as an npm module.

**Comment by mikaelbr** [comment]
@andrewluetgers I added a [PR](https://github.com/andrewluetgers/immstruct/pull/1) for replacing the Map with a already available Map, and re-wrote some functions to show better intent. The thing I thought was making it harder to follow was that _pathListeners was changed as a side-effect inside `getListenerNs`. I'd much rather have pure functions - which only change the system by its return value.

File: src/structure.js (hunk 7 above)
Comment (mikaelbr) [inline src/structure.js]: Very true. I noticed this during some testing as well.

**Comment by andrewluetgers** [comment]
@mikaelbr @Dashed are either of you already or wanting to work on implementing the fix for cursor specific callbacks using cursor sentinel values, if not I will.

**Comment by Dashed** [comment]
@andrewluetgers feel free to go ahead.

File: src/structure.js (hunk 7 above)
Comment (andrewluetgers) [inline src/structure.js]: currently working on correcting this

**Comment by andrewluetgers** [comment]
@Dashed @mikaelbr please take a look at the latest changes for unobserving, THANKS!

**Comment by Dashed** [comment]
@andrewluetgers Sorry. I've afk, and out of the country for a while. Within a week or so, I'll definitely review the changes :)

**Comment by mikaelbr** [comment]
@andrewluetgers Looks good to me. I see you extended the `unobserveAll` API to include "unobserve all but this". Must remember to document that.

I feel comfortable merging this as you've added the test of `unobserve` not leaking to other references, and all other tests pass. Thanks for your great work!

**Comment by andrewluetgers** [comment]
Awesome, THANKS
